src/backend/optimizer/plan/createplan.c

```diff
@@ -9063,4 +9064,148 @@ push_locus_down_after_elide_motion(Plan* plan)
 				plan = plan->lefttree;
 		}
 	}
-}
\ No newline at end of file
+}
+
+/*
+ * Restore Entry locus to SingleQE in the root slice.
+ * This is simply a reverse of push_locus_down_after_elide_motion.
+ * The difference is that it's NOT used when creating a plan but rather
+ * after a plan gets created, it's used to modify the plan in offload_entry_to_qe.
+ */
+static void
+replace_entry_locus_with_singleqe(Plan *plan)
+{
+	while (plan && (plan->locustype == CdbLocusType_Entry))
+	{
+		plan->locustype = CdbLocusType_SingleQE;
+		switch (nodeTag(plan))
+		{
+			case T_Motion:
+				return;
+			case T_Append:
+			{
+				List *subplans = NIL;
+				ListCell *cell;
+				subplans = ((Append *) (plan))->appendplans;
+				foreach(cell, subplans)
+				{
+					replace_entry_locus_with_singleqe(lfirst(cell));
+				}
+				break;
+			}
+			case T_SubqueryScan:
+				plan = ((SubqueryScan *)(plan))->subplan;
+				break;
+			case T_NestLoop:
+			case T_MergeJoin:
+			case T_HashJoin:
+				replace_entry_locus_with_singleqe(plan->righttree);
+				/* FALLTHROUGH */
+			default:
+				plan = plan->lefttree;
```

Comment:
What if there is Subplan node?

Reply:
I can't find a case in regress where there's a Subplan node on QD. is this even possible? to me that sounds a bit weird 🤔

Reply:
```sql
 explain(costs off) select oid, relname, (select count(*) from pg_class a where a.relkind = b.relkind) from pg_class b;
                  QUERY PLAN
-----------------------------------------------
 Seq Scan on pg_class b
   SubPlan 1
     ->  Aggregate
           ->  Seq Scan on pg_class a
                 Filter: (relkind = b.relkind)
 Optimizer: Postgres query optimizer
(6 rows)
```